### PR TITLE
Create interrupt_execution and kill methods on LuaCoroutine

### PR DESCRIPTION
**Pull request by RadiantUwU**
Already documented:

`interrupt_execution()` will throw an error `execution interrupted`

`kill()` will throw an error `execution terminated` and will not stop until the thread exits all the way to the top stack.

### CODE REVIEW

File: src/classes/luaCoroutine.cpp [hunk 1]
```diff
@@ -186,10 +190,29 @@ int LuaCoroutine::luaYield(lua_State *state) {
...
+}
+
+void LuaCoroutine::kill() {
+    if (!done) lua_sethook(tState, terminate_hook, LUA_MASKCOUNT, 1); // force it to run next time
+}
+
+void LuaCoroutine::terminate_hook(lua_State* tState,lua_Debug* dbg) {
+   lua_pushstring(tState,"execution terminated");
+   lua_error(tState);
+}
```
Comment (Trey2k) [inline src/classes/luaCoroutine.cpp:218]: The hook in this case would remain. Meaning if we tried to use the state again it will error out immediately no?

Comment (RadiantUwU) [reply src/classes/luaCoroutine.cpp:218]: Yep, but isnt it hooked only to the running coroutine?

Comment (Trey2k) [reply src/classes/luaCoroutine.cpp:218]: It should be, be what if someone attempts to load another string or file into its state? Normally that should reset it.

Comment (RadiantUwU) [reply src/classes/luaCoroutine.cpp:218]: > It should be, but what if someone attempts to load another string or file into its state? Normally that should reset it.

I'll add the checks then.

**Comment by RadiantUwU** [comment]
Added the checks.

File: src/classes/luaCoroutine.cpp [hunk 2]
```diff
@@ -186,10 +193,32 @@ int LuaCoroutine::luaYield(lua_State *state) {
...
    lua_yield(tState, 0);
 }
+
+void LuaCoroutine::interrupt_execution() {
+    if (!done) lua_sethook(tState, interrupt_hook, LUA_MASKCOUNT, 1); // force it to run next time
+}
+
+void LuaCoroutine::interrupt_hook(lua_State* tState,lua_Debug* dbg) {
+   lua_sethook(tState, nullptr, NULL, NULL); // remove hook
+   lua_pushstring(tState,"execution interrupted");
```
Comment (Trey2k) [inline src/classes/luaCoroutine.cpp:210]:
```suggestion
   lua_pushstring(tState, "execution interrupted");
```

File: src/classes/luaCoroutine.cpp [hunk 3]
```diff
@@ -186,10 +193,32 @@ int LuaCoroutine::luaYield(lua_State *state) {
...
+
+void LuaCoroutine::kill() {
+    if (!done) {
+        lua_sethook(tState, terminate_hook, LUA_MASKCOUNT, 1); // force it to run next time
+        killing = true;
+    }
+}
+
+void LuaCoroutine::terminate_hook(lua_State* tState,lua_Debug* dbg) {
+   lua_pushstring(tState,"execution terminated");
```
Comment (Trey2k) [inline src/classes/luaCoroutine.cpp:222]:
```suggestion
void LuaCoroutine::terminate_hook(lua_State* tState, lua_Debug* dbg) {
   lua_pushstring(tState, "execution terminated");
```

**Review by Trey2k — CHANGES_REQUESTED** [review]
Just some style changes

**Review by Trey2k — APPROVED** [review]

**Comment by RadiantUwU** [comment]
```gdscript
extends Node

var lua: LuaAPI
var coroutine: LuaCoroutine
const killSec := 10


func _ready():
	lua = LuaAPI.new()
	coroutine = lua.new_coroutine()
	coroutine.load_string("
	i = 0
	while true do
		pcall(function() while true do i = i + 1 end end)
	end
	")

var running := false
var timeRunning := 0.0
var thread := Thread.new()
func _process(delta):
	if not running:
		thread.start(_thread)
		running = true
	else:
		timeRunning += delta
		if timeRunning >= 2*killSec and thread.is_started():
			coroutine.kill()
			thread.wait_to_finish()
			timeRunning = 0
		elif timeRunning >=killSec and thread.is_started():
			coroutine.interrupt_execution()
			


func _thread():
	print("Start")
	var ret = coroutine.resume()
	running=false
	if ret is LuaError:
		print("ERROR %d: " % ret.type + ret.msg)
		return
	print("Paused")```

**Comment by Trey2k** [comment]
Tested and confirmed to work with the following script
```gdscript
extends Node

var lua: LuaAPI
var coroutine: LuaCoroutine
const killSec := 10


func _ready():
	lua = LuaAPI.new()
	lua.bind_libraries(["base"])
	coroutine = lua.new_coroutine()
	var err = coroutine.load_string("
	i = 0
	while true do
		pcall(function() while true do i = i + 1 end end)
	end
	")
	if err is LuaError:
		print("ERROR %d: " % err.type + err.message)

var running := false
var timeRunning := 0.0
var thread := Thread.new()
var tried_interrupt := false
func _process(delta):
	if not running:
		thread.start(_thread)
		running = true
	else:
		timeRunning += delta
		if timeRunning >= 2*killSec and thread.is_started():
			coroutine.kill()
			thread.wait_to_finish()
			timeRunning = 0
		elif timeRunning >=killSec and thread.is_started() and not tried_interrupt:
			coroutine.interrupt_execution()
			tried_interrupt = true
			


func _thread():
	print("Start")
	var ret = coroutine.resume()
	running=false
	if ret is LuaError:
		print("ERROR %d: " % ret.type + ret.message)
		return
	print("Killed")
```

**Pull request merged**